// quests/objective.ts
import { QuestType } from './quest-actions';
import { QuestTarget } from './target';
import * as guid from 'guid';

/**
 * A simple contract for a quest objective.
 */
export class QuestObjective {
    // Public properties
    public type: QuestType;
    public target: QuestTarget;
    public count: number;
    public location;
    public id: number;

    constructor(type: QuestType, target: QuestTarget, id: number, count: number = 1, location = null) {
        this.type = type;
        this.target = target;
        this.id = id;
        this.count = count;
        this.location = location;
    }

    serialize(): string {
        return JSON.stringify({
            type: this.type,
            target: this.target.toString(),
            count: this.count,
            location: this.location,
            id: this.id
        });
    }

    static deserialize(input: string): QuestObjective {
        var data = JSON.parse(input);
        return new QuestObjective(data.type, data.target, data.id, data.count, data.location);
    }
}

/**
 * Represents a completion status for a quest objective.
 */
export class QuestObjectiveCompletionStatus {
    obj: QuestObjective;
    count: number;

    private _eventHandlers: {};
    private _cancelToken;
    private _cancelReason: string;
    private _isWatching: boolean = false;

    constructor(obj: QuestObjective, count: number = 0) {
        this.obj = obj;
        this.count = count;
        this._eventHandlers = {
            "$progress": [],
            "$cancelled": [],
            "$completed": []
        };
    }

    isComplete() {
        return this.count == this.obj.count;
    }

    /**
     * Registers an event handler to be executed.
     * @param event 
     * @param handler 
     */
    on(event: string, handler: (QuestObjective, EventArgs) => void) {
        if (!this._eventHandlers[event]) throw new Error('Event "' + event + '" does not exist.');
        this._eventHandlers[event].push(handler);
    }

    private callHandlersFor(property, eventArgs) {
        var handlers = this._eventHandlers[property];
        for (let i = 0; i < handlers.length; i++) {
            handlers[i](this, eventArgs);
        }
    }

    private finalize(e) {
        this.callHandlersFor('$completed', e);
        this._cancelToken.unregister();
    }

    /**
     * Registers the events to watch.
     */
    beginWatch(player) {
        if (this._isWatching || this.isComplete()) return;
        this._isWatching = true;
        if (this.count <= this.count) {
            this.callHandlersFor('$completed', undefined);
            return;
        }
        if (this.obj.type == QuestType.BREAK) {
            this._cancelToken = 
            eventHandler('block', 'break', (e) => {
                if (player.getUniqueId() == e.player.getUniqueId() && this.obj.target == e.getBlock().type) {
                    this.count++;
                    this.callHandlersFor('$progress', e);
                }
                if (this.count <= this.count) {
                    this.finalize(e);
                }
            });
        }
        if (this.obj.type == QuestType.BREED) {
            this._cancelToken =
            eventHandler('entity', 'breed', (e) => {
                if (e.getBreeder().getUniqueId != undefined &&
                    e.getBreeder().getUniqueId() == player.getUniqueId() &&
                    e.getEntityType == this.obj.target) {
                        this.count++;
                        this.callHandlersFor('$progress', e);
                    }
                if (this.count <= this.count) {
                    this.finalize(e);
                }
            });
        }
        if (this.obj.type == QuestType.COLLECT) {
            this._cancelToken = 
            eventHandler('inventory', 'pickupItem', (e) => {
                if (e.getInventory().class == org.bukkit.inventory.PlayerInventory.class &&
                    e.getInventory().getHolder != undefined &&
                    e.getInventory().getHolder().getUniqueId() == player.getUniqueId() &&
                    e.getItem().getItemStack().type == this.obj.target) {
                        this.count += e.getItem().getItemStack().getAmount();
                        this.callHandlersFor('$progress', e);
                    }
                if (this.count <= this.count) {
                    this.finalize(e);
                }
            });
        }
        if (this.obj.type == QuestType.CRAFT) {
            this._cancelToken =
            eventHandler('inventory', 'craft', (e) => {
                if (e.getInventory().class == org.bukkit.inventory.PlayerInventory.class &&
                    e.getInventory().getHolder != undefined &&
                    e.getInventory().getHolder().getUniqueId() == player.getUniqueId() &&
                    e.getRecipe().getResult().type == this.obj.target) {
                        this.count += e.getRecipe().getResult().getAmount();
                        this.callHandlersFor('$progress', e);
                    }
                if (this.count <= this.count) {
                    this.finalize(e);
                }
            });
        }
        if (this.obj.type == QuestType.FISH) {
            this._cancelToken =
            eventHandler('player', 'fish', (e) => {
                if (e.getPlayer().getUniqueId() == player.getUniqueId()) {
                    this.count++;
                    this.callHandlersFor('$progress', e);
                }
                if (this.count <= this.count) {
                    this.finalize(e);
                }
            });
        }
        if (this.obj.type == QuestType.KILL) {
            this._cancelToken = 
            eventHandler('entity', 'death', (e) => {
                if (e.getEntity().getKiller() != null &&
                    e.getEntity().getKiller().getUniqueId != undefined && 
                    e.getEntity().getKiller().getUniqueId() == player.getUniqueId() &&
                    e.getEntity().getType() == this.obj.target) {
                        this.count++;
                        this.callHandlersFor('$progress', e);
                    }
                if (this.count <= this.count) {
                    this.finalize(e);
                }
            });
        }
        if (this.obj.type == QuestType.LOCATE) {
            this._cancelToken =
            eventHandler('player', 'move', (e) => {
                if (e.getTo().distanceSquared(this.obj.target) < 10) {
                    this.finalize(e);
                }
            });
        }
        if (this.obj.type == QuestType.PLACE) {
            this._cancelToken =
            eventHandler('block', 'place', (e) => {
                if (e.getPlayer().getUniqueId() ==  player.getUniqueId() && 
                    e.getBlock().type == this.obj.target.type) {
                        this.finalize(e);
                    }
            });
        }
        if (this.obj.type == QuestType.SMELT) {
            this._cancelToken = 
            eventHandler('inventory', 'extract', (e) => {
                if (e.getPlayer().getUniqueId() == player.getUniqueId() && 
                    e.getItemType() == this.obj.target) {
                        this.count += e.getItemAmount();
                        this.callHandlersFor('$progress', e);
                    }
                if (this.count <= this.count) {
                    this.finalize(e);
                }
            });
        }
    }

    toString() {
        return `${this.obj.type.toString().toLowerCase()} ${this.count} ${this.obj.target.toString().toLowerCase().replace('_', ' ')}`;
    }
}
